File: src/backend/agentchat/api/v1/knowledge.py
from typing import Union, List
from loguru import logger
from fastapi import Body, APIRouter, Depends

from agentchat.api.services.knowledge import KnowledgeService
from agentchat.api.services.user import get_login_user, UserPayload
from agentchat.schema.knowledge import KnowledgeCreateRequest, KnowledgeUpdateRequest
from agentchat.schema.schemas import UnifiedResponseModel, resp_500, resp_200
from agentchat.services.rag_handler import RagHandler

router = APIRouter(tags=["Knowledge"])


@router.post("/knowledge/create", response_model=UnifiedResponseModel)
async def upload_knowledge(*,
                           knowledge_req: KnowledgeCreateRequest,
                           login_user: UserPayload = Depends(get_login_user)):
    try:
        await KnowledgeService.create_knowledge(knowledge_req.knowledge_name, knowledge_req.knowledge_desc,
                                                login_user.user_id)
        return resp_200()
    except Exception as err:
        return resp_500(message=str(err))


@router.get("/knowledge/select", response_model=UnifiedResponseModel)
async def select_knowledge(login_user: UserPayload = Depends(get_login_user)):
    try:
        results = await KnowledgeService.select_knowledge(login_user.user_id)
        return resp_200(data=results)
    except Exception as err:
        logger.error(err)
        return resp_500(message=str(err))


@router.put("/knowledge/update", response_model=UnifiedResponseModel)
async def update_knowledge(*,
                           knowledge_req: KnowledgeUpdateRequest,
                           login_user: UserPayload = Depends(get_login_user)):
    try:
        # 验证用户权限
        await KnowledgeService.verify_user_permission(knowledge_req.knowledge_id, login_user.user_id)

        await KnowledgeService.update_knowledge(knowledge_req.knowledge_id, knowledge_req.knowledge_name,
                                                knowledge_req.knowledge_desc)
        return resp_200()
    except Exception as err:
        logger.error(err)
        return resp_500(message=str(err))


@router.delete("/knowledge/delete", response_model=UnifiedResponseModel)
async def delete_knowledge(knowledge_id: str = Body(embed=True),
                           login_user: UserPayload = Depends(get_login_user)):
    try:
        # 验证用户权限
        await KnowledgeService.verify_user_permission(knowledge_id, login_user.user_id)

        await KnowledgeService.delete_knowledge(knowledge_id)
        return resp_200()
    except Exception as err:
        logger.error(err)
        return resp_500(message=str(err))

@router.post("/knowledge/retrieval", response_model=UnifiedResponseModel)
async def retrieval_knowledge(*,
                              query: str = Body(..., description="用户的问题"),
                              knowledge_id: Union[str, List[str]] = Body(..., description="知识库ID")):
    if isinstance(knowledge_id, str):
        content = await RagHandler.retrieve_ranked_documents(query, [knowledge_id], [knowledge_id])
    else:
        content = await RagHandler.retrieve_ranked_documents(query, knowledge_id, knowledge_id)
    return resp_200(content)

File: src/frontend/src/apis/knowledge.ts
import { request } from "../utils/request"

// 统一响应模型
export interface UnifiedResponse<T = any> {
  status_code: number
  status_message: string
  data?: T
}

// 知识库响应类型
export interface KnowledgeResponse {
  id: string
  name: string
  description: string | null
  user_id: string | null
  create_time: string
  update_time: string
  count: number // 文件数量
  file_size: string // 文件总大小（已格式化）
}

// 知识库创建请求
export interface KnowledgeCreateRequest {
  knowledge_name: string
  knowledge_desc?: string
}

// 知识库更新请求
export interface KnowledgeUpdateRequest {
  knowledge_id: string
  knowledge_name?: string
  knowledge_desc?: string
}

// 知识库删除请求
export interface KnowledgeDeleteRequest {
  knowledge_id: string
}

// 知识库检索请求
export interface KnowledgeRetrievalRequest {
  query: string
  knowledge_id: string | string[]
  top_k?: number
}

// 查询知识库列表
export function getKnowledgeListAPI() {
  return request<UnifiedResponse<KnowledgeResponse[]>>({
    url: '/api/v1/knowledge/select',
    method: 'GET'
  })
}

// 创建知识库
export function createKnowledgeAPI(data: KnowledgeCreateRequest) {
  return request<UnifiedResponse<null>>({
    url: '/api/v1/knowledge/create',
    method: 'POST',
    data
  })
}

// 更新知识库
export function updateKnowledgeAPI(data: KnowledgeUpdateRequest) {
  return request<UnifiedResponse<null>>({
    url: '/api/v1/knowledge/update',
    method: 'PUT',
    data
  })
}

// 删除知识库
export function deleteKnowledgeAPI(data: KnowledgeDeleteRequest) {
  return request<UnifiedResponse<null>>({
    url: '/api/v1/knowledge/delete',
    method: 'DELETE',
    data
  })
}

// 知识库检索
export function knowledgeRetrievalAPI(data: KnowledgeRetrievalRequest) {
  return request<UnifiedResponse<string>>({
    url: '/api/v1/knowledge/retrieval',
    method: 'POST',
    data
  })
} 